### ISSUE TITLE
Cloudflare Turnstile fails (true domain leaked?) ("Verifying you are human" page)

### ISSUE THREAD

**Issue by itschasa** [titaniumnetwork-dev/Ultraviolet#121]
will always get stuck on this page:

![image](https://github.com/titaniumnetwork-dev/Ultraviolet/assets/79016507/4f1e5bca-da3c-4b77-a375-3dfa16311da3)

the captcha doesnt load, if anyone has any other results, please share.

**Comment by Percslol** [comment]
known bug, will keep this issue open for discussion and possible fixes

**Comment by itschasa** [comment]
seems like turnstile still works whilst inside an iframe, so this could be a fixable issue, however, is likely to be a cat and mouse game with captcha providers

![image](https://github.com/titaniumnetwork-dev/Ultraviolet/assets/79016507/0e406d7e-f6c8-4624-b1d2-e4148b26740c)

**Comment by madeline-yana** [comment]
I think this is an issue if you are deploying via a known big server provider (e.g. if the provider uses Hetzner), Cloudflare probably just blocks the IPs.

**Comment by itschasa** [comment]
i see this in console when a captcha is attempted:
`[Cloudflare Turnstile] Ignored message from wrong origin: https://*site with captcha*.`
![image](https://github.com/titaniumnetwork-dev/Ultraviolet/assets/79016507/cf7a168b-2c7e-4261-8f59-2e648b05090d)


maybe turnstile is getting the actual domain of the proxy, and using that to check [Message](https://developer.mozilla.org/en-US/docs/Web/API/Window/message_event) events follow the corrent origin as part of a check? (or just because it uses [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postmessage) to do the challenge)

this might be solvable, if we can find how turnstile is getting the true domain of the proxy

**Comment by Percslol** [comment]
> i see this in console when a captcha is attempted: `[Cloudflare Turnstile] Ignored message from wrong origin: https://*site with captcha*.` ![image](https://private-user-images.githubusercontent.com/79016507/325677395-cf7a168b-2c7e-4261-8f59-2e648b05090d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwOTAxOTksIm5iZiI6MTcxNDA4OTg5OSwicGF0aCI6Ii83OTAxNjUwNy8zMjU2NzczOTUtY2Y3YTE2OGItMmM3ZS00MjYxLThmNTktMmU2NDhiMDUwOTBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI2VDAwMDQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlOTg0MmQzZjFmZjIwZmE3NWFlODM4MDg3YTRlMjcyOWZjMGFmNTIwYjZhMTYwOGU2NzM1YjM2ZGE3OWY5NzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4qwI8H7lT8qrv670hgfYuG7FO1upT19N70TYToq-hc0)
> 
> maybe turnstile is getting the actual domain of the proxy, and using that to check [Message](https://developer.mozilla.org/en-US/docs/Web/API/Window/message_event) events follow the corrent origin as part of a check? (or just because it uses [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postmessage) to do the challenge)
> 
> this might be solvable, if we can find how turnstile is getting the true domain of the proxy

yeah i just checked this, seems to be reproducible

fixing this might also solve #128

**Comment by Percslol** [comment]
yeah the iframe is sending the wrong location in the post message

![image](https://github.com/titaniumnetwork-dev/Ultraviolet/assets/83934299/b39a5284-89d7-4b37-94c0-d37aec8ae273)

**Comment by Percslol** [comment]
![image](https://github.com/titaniumnetwork-dev/Ultraviolet/assets/83934299/f03cbe14-cf52-40c7-806e-966677c5d89e)

**Comment by itschasa** [comment]
> yeah the iframe is sending the wrong location in the post message
> 
> 
> 
> ![image](https://github.com/titaniumnetwork-dev/Ultraviolet/assets/83934299/b39a5284-89d7-4b37-94c0-d37aec8ae273)
> 
> 

would this be due to a bad rewrite of the captcha's js? they are heavily obfuscated so it'd make sense

**Comment by itschasa** [comment]
overriding the origin might need to be changed

https://github.com/titaniumnetwork-dev/Ultraviolet/blob/1d38959da170abb1c195187100aef5134967e023/src/client/message.js#L105

**Comment by marriedtopython** [comment]
**This is a fix I did. It might not work for you, but it is working for me**

```
    referer: "https://www.google.com", // Spoof referer
    headers: {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        "Referer": "https://www.google.com/",
        "Origin": "https://www.google.com"
    }

};

// fixed uv 
```

**Comment by itschasa** [comment]
does cloudflare just blindly accept the origin of messages if it's google.com?

**Comment by marriedtopython** [comment]
> does cloudflare just blindly accept the origin of messages if it's google.com?

Not really, it's just a placeholder, it can work theoretically. Usually I use the url that I use for the origin etc
